### PR TITLE
Dynamic color detector moved to own class

### LINKED ISSUE #10: create a separate class for dynamic colorspaces
The PixelListColorDetector should not be dynamic. But another new detector inheriting from it should.

### PR DESCRIPTION
**Pull request by JanGut**
This moves the implementation of the DynamicPixelListColorDetector to an own class inheriting from the static PixelListColorDetector.

Closes #10.

### CODE REVIEW

File: bitbots_vision/src/bitbots_vision/dynamic_color_space.py [hunk 1]
```diff
@@ -69,25 +68,24 @@ def vision_config_callback(self, msg):
         :return: None
         """
         # Load dict from string in yaml-format in msg.data
-        vision_config = yaml.load(msg.data)
```
Comment (JanGut) [inline bitbots_vision/src/bitbots_vision/dynamic_color_space.py]: yaml.load is deprecated

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 2]
```diff
@@ -284,22 +275,92 @@ def mask_image(self, image):
...
+        mask = VisionExtensions.maskImg(image, self.color_space)
+
+        # toggle publishing of 'field_mask'-messages   
+        if self.config['vision_field_mask_image']:
+            self.imagepublisher.publish(self.bridge.cv2_to_imgmsg(mask, '8UC1'))
+
+        return mask
+
+
+class DynamicPixelListColorDetector(PixelListColorDetector):
```
Comment (JanGut) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py]: This needs to be tested on a robot

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 3]
```diff
@@ -13,13 +13,17 @@
...
+    or to create masked binary images.
+    """
+
     def __init__(self, debug_printer):
         # type: (DebugPrinter) -> None
         """
-        ColorDetector is abstract super-class of specialized sub-classes.
-        ColorDetectors are used e.g. to check, if a pixel matches the defined color space
-        or to create masked binary images.
+        Initiating of ColorDetector.
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py]: Initialization of ColorDetector

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 4]
```diff
@@ -94,18 +98,21 @@ def pixel_bgr2hsv(pixel):
...
+    """
+    HsvSpaceColorDetector is a ColorDetector, that is based on the HSV-color space.
+    The HSV-color space is adjustable by setting min- and max-values for hue, saturation and value.
+    """
     def __init__(self, debug_printer, min_vals, max_vals):
         # type: (DebugPrinter, tuple[int, int, int], tuple[int, int, int]) -> None
         """
-        HsvSpaceColorDetector is a ColorDetector, that is based on the HSV-color space.
-        The HSV-color space is adjustable by setting min- and max-values for hue, saturation and value.
+        Initiating of HsvSpaceColorDetector.
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py]: Initialization of...

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 5]
```diff
@@ -173,64 +180,48 @@ def mask_image(self, image):
...
+    """
+
+    def __init__(self, debug_printer, package_path, config):
         # type:(DebugPrinter, str, dict, bool) -> None
         """
-        PixelListColorDetector is a ColorDetector, that is based on a color space.
-        The color space is initially loaded from color-space-file at color_path (in config)
-        and optionally adjustable to changing color conditions (dynamic color space).
-        The color space is represented by boolean-values for RGB-color-values.
+        Initiating of PixelListColorDetector.
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py]: Initialization of...

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 6]
```diff
@@ -173,64 +180,48 @@ def mask_image(self, image):
...
         # Register publisher for 'mask_image'-messages
         self.imagepublisher = rospy.Publisher(
-            "/mask_image",
-            Image,
-            queue_size=1)
-
-        # Register publisher for 'mask_image_dyn'-messages
-        self.imagepublisher_dyn = rospy.Publisher(
-            "/mask_image_dyn",
+            "/field_mask",
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py]: this topic could be configurable

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 7]
```diff
@@ -284,22 +275,92 @@ def mask_image(self, image):
...
+        
+        # toggle publishing of mask_img_dyn msg with dynamic color space
+        self.publish_dyn_field_mask_msg = self.config['dynamic_color_space_field_mask_image']
+
+        # toggle use of dynamic color space
+        self.dynamic_color_space_turned_on = self.config['dynamic_color_space']
+
+        # Subscribe to 'color_space'-messages from DynamicColorSpace
+        self.color_space_subscriber = rospy.Subscriber(
+            'color_space',
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py]: This HAS to be configurable.

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 8]
```diff
@@ -284,22 +275,92 @@ def mask_image(self, image):
...
+            ColorSpace,
+            self.color_space_callback,
+            queue_size=1,
+            buff_size=2**20)
+
+        # Register publisher for 'mask_image_dyn'-messages
+        self.imagepublisher_dyn = rospy.Publisher(
+            "/dynamic_field_mask",
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py]: This should be configurable.

File: bitbots_vision/cfg/vision.cfg [hunk 9]
```diff
@@ -81,7 +81,7 @@ group_horizon_finder.add("horizon_finder_min_precision_pix", int_t, 0, "horizon_
 
 group_vision.add("vision_debug_image", bool_t, 0, "vision_debug_image")
 group_vision.add("vision_debug_image_msg", bool_t, 0, "vision_debug_image_msg")
-group_vision.add("vision_mask_img_msg", bool_t, 0, "vision_mask_img_msg")
+group_vision.add("vision_field_mask_image", bool_t, 0, "Publish field mask image message for debug purposes")
```
[the remark this answers is not in the document]
Comment (JanGut) [reply bitbots_vision/cfg/vision.cfg]: closed with d0589806e79f9db07bc905cdd51e47b25377c253

File: bitbots_vision/cfg/vision.cfg [hunk 10]
```diff
@@ -115,8 +115,8 @@ group_ball_finder.add("ball_finder_min_size", int_t, 0, "ball_finder_min_size",
 group_classifier_model.add("classifier_model_path", str_t, 0, "classifier_model_path", None)
 group_cascade_classifier.add("cascade_classifier_path", str_t, 0, "cascade_classifier_path", None)
 
-group_dynamic_color_space.add("dynamic_color_space", bool_t, 0, "vision_dynamic_color_space")
-group_dynamic_color_space.add("dynamic_color_space_mask_img_dyn_msg", bool_t, 0, "vision_dynamic_color_space_mask_img_dyn_msg")
+group_dynamic_color_space.add("dynamic_color_space", bool_t, 0, "Turn dynamic color space ON or OFF")
+group_dynamic_color_space.add("dynamic_color_space_field_mask_image", bool_t, 0, "Publish dynamic field mask image message for debug purposes")
```
[the remark this answers is not in the document]
Comment (JanGut) [reply bitbots_vision/cfg/vision.cfg]: closed with d0589806e79f9db07bc905cdd51e47b25377c253

File: bitbots_vision/cfg/vision.cfg [hunk 11]
```diff
@@ -115,8 +115,8 @@ group_ball_finder.add("ball_finder_min_size", int_t, 0, "ball_finder_min_size",
 group_classifier_model.add("classifier_model_path", str_t, 0, "classifier_model_path", None)
 group_cascade_classifier.add("cascade_classifier_path", str_t, 0, "cascade_classifier_path", None)
 
-group_dynamic_color_space.add("dynamic_color_space", bool_t, 0, "vision_dynamic_color_space")
-group_dynamic_color_space.add("dynamic_color_space_mask_img_dyn_msg", bool_t, 0, "vision_dynamic_color_space_mask_img_dyn_msg")
+group_dynamic_color_space.add("dynamic_color_space", bool_t, 0, "Turn dynamic color space ON or OFF")
```
[the remark this answers is not in the document]
Comment (JanGut) [reply bitbots_vision/cfg/vision.cfg]: closed with d0589806e79f9db07bc905cdd51e47b25377c253

File: bitbots_vision/src/bitbots_vision/vision.py [hunk 12]
```diff
@@ -274,11 +282,17 @@ def _dynamic_reconfigure_callback(self, config, level):
             [config['blue_color_detector_upper_values_h'], config['blue_color_detector_upper_values_s'],
              config['blue_color_detector_upper_values_v']])
 
-        self.field_color_detector = color.PixelListColorDetector(
-            self.debug_printer,
-            self.package_path,
-            config,
-            primary_detector=True)
+        if config['dynamic_color_space_active']:
```
Comment (Flova) [inline bitbots_vision/src/bitbots_vision/vision.py]: Redundant deactivation of the dynamic color space using this config parameter.

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py (hunk 3 above)
Comment (JanGut) [reply bitbots_vision/src/bitbots_vision/vision_modules/color.py]: closed with 7398d9f2ebe6647e5caa15e9846bfb3e33624e37

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py (hunk 4 above)
Comment (JanGut) [reply bitbots_vision/src/bitbots_vision/vision_modules/color.py]: Closed with 7398d9f2ebe6647e5caa15e9846bfb3e33624e37

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py (hunk 5 above)
Comment (JanGut) [reply bitbots_vision/src/bitbots_vision/vision_modules/color.py]: Closed with 7398d9f2ebe6647e5caa15e9846bfb3e33624e37

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 13]
```diff
@@ -284,22 +275,92 @@ def mask_image(self, image):
...
+            if (self.primary_detector and self.publish_dyn_field_mask_msg):
                 self.imagepublisher_dyn.publish(self.bridge.cv2_to_imgmsg(dyn_mask, '8UC1'))
   
-        # toggle publishing of 'mask_img'-messages          
-        if (self.primary_detector and self.publish_mask_img_msg):
+        # toggle publishing of 'field_mask'-messages          
+        if (self.primary_detector and self.publish_field_mask_img_msg):
             self.imagepublisher.publish(self.bridge.cv2_to_imgmsg(static_mask, '8UC1'))
 
         if self.dynamic_color_space_turned_on:
```
Comment (Flova) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py]: See my other comment.

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py (hunk 6 above)
Comment (JanGut) [reply bitbots_vision/src/bitbots_vision/vision_modules/color.py]: closed with 1d1bfc5babf81c18a5b762e4508d002de9d94d5c and c7c947b904fc75ac6834c9e8ae846a67bbeb3df5

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py (hunk 7 above)
Comment (JanGut) [reply bitbots_vision/src/bitbots_vision/vision_modules/color.py]: closed with 1d1bfc5babf81c18a5b762e4508d002de9d94d5c and c7c947b904fc75ac6834c9e8ae846a67bbeb3df5

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py (hunk 8 above)
Comment (JanGut) [reply bitbots_vision/src/bitbots_vision/vision_modules/color.py]: closed with 1d1bfc5babf81c18a5b762e4508d002de9d94d5c and c7c947b904fc75ac6834c9e8ae846a67bbeb3df5

File: bitbots_vision/src/bitbots_vision/vision.py (hunk 12 above)
Comment (JanGut) [reply bitbots_vision/src/bitbots_vision/vision.py]: closed with 39a9cd0

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py (hunk 13 above)
Comment (JanGut) [reply bitbots_vision/src/bitbots_vision/vision_modules/color.py]: closed with 39a9cd0

File: bitbots_vision/src/bitbots_vision/dynamic_color_space.py (hunk 1 above)
Comment (JanGut) [reply bitbots_vision/src/bitbots_vision/dynamic_color_space.py]: For PyYAML-3.1 yaml.safe_load() is the safe method to load yaml.
Closed with ab393636e3e042fde61a03fd444f1b4133c182d6.

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py (hunk 2 above)
Comment (JanGut) [reply bitbots_vision/src/bitbots_vision/vision_modules/color.py]: Tested on standalone jetson and running...

**Pull request merged**